### PR TITLE
Use itertools instead of list concatenation, do not use bare except

### PR DESCRIPTION
**Pull request by sarum9in**

### CODE REVIEW

File: mesonbuild/scripts/meson_install.py [hunk 1]
```diff
@@ -283,7 +283,7 @@ def run_install_script(d):
             rc = subprocess.call(script + args, env=child_env)
             if rc != 0:
                 sys.exit(rc)
-        except:
+        except Exception:
```
Comment (TingPing) [inline mesonbuild/scripts/meson_install.py]: This probably should have been `subprocess.check_call()` and `subprocess.CalledProcessError`

File: test cases/common/72 build always/version_gen.py [hunk 2]
```diff
@@ -12,7 +12,7 @@ def generate(infile, outfile, fallback):
         (stdo, _) = p.communicate()
         if p.returncode == 0:
             version = stdo.decode().strip()
-    except:
+    except Exception:
```
Comment (TingPing) [inline test cases/common/72 build always/version_gen.py]: Should maybe change this to use `subprocess.check_output()` and handle `subprocess.CalledProcessError`?

**Comment by sarum9in** [comment]
@TingPing thanks for feedback, I will take a closer look on this.

**Comment by infirit** [comment]
What kind of speed increase did you see with itertools.chain? I suspect the performance increase  is negligible on two small lists.

**Comment by sarum9in** [comment]
@infirit  There is no speed increase, nothing in this patch indicates that performance improvement is the goal. But `itertools.chain()` usage is more consistent with other code and avoids unnecessary copy of lists.

**Comment by nirbheek** [comment]
subprocess.check_call can fail in five ways (that I know of): subprocess.CalledProcessError, WindowsError, OSError, PermissionError, FileNotFoundError. We may want to store this as a tuple in mesonlib and use `except that_tuple:`.

**Comment by sarum9in** [comment]
@nirbheek that makes me think that we don't want this and leaving `Exception` is fine. It will not match `KeyoardInterrupt` or `SystemExit` exceptions which should not be caught, but will be simple enough to match whatever it throws.

**Comment by QuLogic** [comment]
> WindowsError, OSError, PermissionError, FileNotFoundError

These are all subclasses of `OSError`, so it's not like you need to catch all of them.

**Comment by TingPing** [comment]
Also the entire point of that flake warning is because broad exceptions can have unintended side-effects, so just replacing it with `Exception` is only hiding the problem.

**Comment by sarum9in** [comment]
@TingPing well as I understand the issue with bare `except` is that you may catch special exception classes also, which are not inherited from Exception. But if we can catch something specific like `OSError` it's fine as well. I just didn't have time to take a look at this today :) sorry

**Comment by sarum9in** [comment]
Looking closer at the code in question, it can also raise `UnicodeDecodeError` if `git describe` returns something obscure. In my opinion what we care about here is that if anything goes wrong we should fallback, otherwise proceed. We specifically don't want to catch sys.exit() or interrupt, but everything else we want. I don't think we are hiding anything by using `Exception` because this is intention. Listing types explicitly however exposes us to possibility to forget something, like we almost did now. So I am going to leave it as `Exception`.

Ready for review.

**Comment by nirbheek** [comment]
> We specifically don't want to catch sys.exit() or interrupt, but everything else we want. I don't think we are hiding anything by using Exception because this is intention

`Exception` also catches `SyntaxError`, `TypeError`, and other unexpected errors. We don't really want to do that here or elsewhere. It might sound innocuous now, but it's going to bite us eventually.

> These are all subclasses of OSError, so it's not like you need to catch all of them.

I didn't know this, so this seems like the best option to me right now: `(subprocess.CalledProcessError, OSError)`. Of course, it's up to @jpakkane so let's just do whatever he says.

**Comment by sarum9in** [comment]
`SyntaxError` and friends is a good argument against `Exception`, I agree with it.

**Pull request merged**